Fix typed property without default value support

sspat opened the pull request:
Added testcase for bug described in issue #7854

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH7854Test.php:
```diff
@@ -0,0 +1,52 @@
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+/**
+ * @group GH7854
```
Please add [`@requires PHP >= 7.4`](https://phpunit.readthedocs.io/en/8.4/incomplete-and-skipped-tests.html#skipping-tests-using-requires) to make sure we don't run this test for other PHP versions.

lcobucci replied:
@Ocramius do you have any suggestion on how to avoid syntax errors for PHP < 7.3 here? The only thing I can think of is moving the classes to an external file or using `eval()` :nauseated_face:

Ocramius replied:
I wouldn't mind bumping the dependency :trollface:

greg0ire replied:
I just had a look at [the PHPUnit docs](https://phpunit.readthedocs.io/en/8.4/incomplete-and-skipped-tests.html#skipping-tests-using-requires)

> If you are using syntax that doesn’t compile with a certain PHP Version look into the xml configuration for version dependent includes in The <testsuites> Element

Following the link leads to reading about a way to add a directory to a test suite only when matching some constraints:

```
<testsuites>
  <testsuite name="unit">
    <directory phpVersion="8.0.0" phpVersionOperator=">=">tests/unit</directory>
  </testsuite>
</testsuites>
```

greg0ire replied:
I like the first solution better, only one directory to create.

greg0ire replied:
Wait the tests pass now, I thought there were an issue with PHP < 7.3 as @lcobucci mentioned. Why is it not the case?

greg0ire replied:
Ah that's because the class was defined in the same file at first, isn't it?

sspat replied:
Yes, then I moved it to a separate file.

greg0ire replied:
So I guess we should now decide if we stay with the separate file or if we go with the solution I described above. @lcobucci , @Ocramius , any opinions?

lcobucci replied:
Didn't know about that PHPUnit feature :scream: It's much better than having a separate file. @sspat sorry about the back and forth but could you change it? :grimacing:

sspat replied:
@lcobucci no problem, done!

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH7854Test.php:
```diff
@@ -0,0 +1,52 @@
+class GH7854Entity
+{
+    /**
+     * @Embedded(class = "GH7854ValueObject")
```
This should use `::class` syntax instead

```suggestion
     * @Embedded(class=GH7854ValueObject::class)
```

lcobucci reviewed (changes requested):
@sspat thanks for this test! A few things we need to consider apart from the fix :+1:

sspat commented:
@lcobucci added suggested changes, also made my try on implementing the fix for this issue.
However, there are a lot of places elsewhere in the codebase where reflection is used and the same issue could arise. If my implementation is going in the right direction, I could investigate further.

greg0ire commented:
Please associate your email address with your Github account, or change the email in your commits to an address already associated with it.

sspat commented:
> Please associate your email address with your Github account, or change the email in your commits to an address already associated with it.

Fixed, thank you for pointing this out!

SenseException commented:
I've restarted the Scrutinizer build. Checks are all green now.

greg0ire commented on tests/Doctrine/PHP_7_4/Tests/ORM/Functional/Ticket/GH7854Test.php:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * @group GH7854
+ * @requires PHP >= 7.4
```
Is this still needed?

sspat replied:
Probably not, removed id

sspat committed:
Configure suite to have PHP 7.4 specific tests

sspat committed:
Add test case using typed properties in embeddables

sspat committed:
Check if property is initialized to get its value

Preventing the error "Typed property $CLASS::$PROPERTY must not be
accessed before initialization".

lcobucci commented:
@sspat I've just rebased your branch to sync things and rearrange the commits. My hope was to ship this in v2.7.0 but, as you said, we should cover other scenarios not only with embeddable but entity properties as well.

Let's aim for v2.7.1 :+1:

sspat commented:
> @sspat I've just rebased your branch to sync things and rearrange the commits. My hope was to ship this in v2.7.0 but, as you said, we should cover other scenarios not only with embeddable but entity properties as well.
> 
> Let's aim for v2.7.1 👍

Extended the existing test-case, now it contains:
- Typed embeddable value object
- Typed relation with other entity
- Typed entity propety

These all work without any issues.

Also did a search in `/lib/Doctrine/ORM/Mapping` namespace for all instances of `->getValue(`

Apart from `\Doctrine\ORM\Mapping\ReflectionEmbeddedProperty` which is fixed, it is used only in 
`\Doctrine\ORM\Mapping\ClassMetadataInfo::getFieldValue` and `\Doctrine\ORM\Mapping\ClassMetadataInfo::getIdentifierValues`  - both of which use `\Doctrine\Common\Reflection\RuntimePublicReflectionProperty` for reflection which has this problem with accessing uninitialized properties fixed too, but in another way.

Instead of calling `\ReflectionProperty::isInitialized` like I did in `\Doctrine\ORM\Mapping\ReflectionEmbeddedProperty` and `\Doctrine\ORM\UnitOfWork`:
```php
if (method_exists($this->parentProperty, 'isInitialized') && $this->parentProperty->getDeclaringClass()->isInstance($object)) {
    return $this->parentProperty->isInitialized($object) ? $this->parentProperty->getValue($object) : null;
}
```
It does the following:
```php
return isset($object->$name) ? parent::getValue($object) : null;
```
When checking with `isset` on a non-initialized typed property it returns false and no `Error: Typed property must not be accessed before initialization` error is thrown.

Should I commit this extended test?
ShouId I replace my previous solution with `isset`?
Should I continue investigating outside of the `Doctrine\ORM\Mapping` namespace?

beberlei commented:
:-1: from me because of the performance implications. Refleciton setValue/getvalue are in the high performance path. This should be fixed with inheritance by creating different instances for PHP 7.4 if the property is typed, otherwise keep old code.

beberlei commented:
This looks like it could be complicated to fix :) It needs additional class in doctrine/reflection, a change to RuntimeReflectionService in doctrine/persistence so that getAccessibleProperty returns the correct reflection property.

sspat committed:
Reverted implementation, updated test #7854

sspat commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -631,7 +632,11 @@ public function computeChangeSet(ClassMetadata $class, $entity)
 
         foreach ($class->reflFields as $name => $refProp) {
-            $value = $refProp->getValue($entity);
+            if (method_exists($refProp, 'isInitialized') && $refProp->getDeclaringClass()->isInstance($entity)) {
```
@beberlei Thank you very much for the review, it gave me a lot of missing insight.

The current state of the test for this issue seems to be broken, as the issue does not arise for public properties, because for them an instance of `\Doctrine\Common\Reflection\RuntimePublicReflectionProperty` is created, which is checking the object property with `isset` prior to accessing it's value:
```php
return isset($object->$name) ? parent::getValue($object) : null;
```
So the error happens only for protected and private non-initialized properties without a default value.

I reverted my previous fix and updated the test accordingly, it is again failing.

If I understand correctly, we need to first update doctrine/reflection, then update the dependency in doctrine/persistence and implement the fix there, and then update the dependencies in doctrine/orm and the test will pass.

I created https://github.com/doctrine/reflection/pull/22 in doctrine/reflection to fix some PHP 7.4 issues that currently block me from running the tests, then I will proceed with the implementation.

sspat committed:
Fixed paths in phpunit config files for travis

semaex commented:
How is this issue? I have the same problem and don't know if it is already fixed. Thanxs

Metabor commented:
@semaex My Workaround ist to convert the property to allow null (even if it will never be after loading) and set default to null.

So like this
private ?\DateTime $date = null;

kuraobi commented:
@Metabor that wasn't really an option for us, because then we would have to add checks everywhere to ensure the value is not null, otherwise PHPStan throws errors. So for now we are using https://github.com/doctrine/orm/pull/7950 until an official fix is released. It's been working fine in production for 10 days now, although it may not cover all the use cases.

semaex commented:
> @sspat I've just rebased your branch to sync things and rearrange the commits. My hope was to ship this in v2.7.0 but, as you said, we should cover other scenarios not only with embeddable but entity properties as well.
> 
> Let's aim for v2.7.1 👍

Hi, when will be a 2.7.1 release? I'm so crazy with this bug. Thanxs!

beberlei commented:
No 2.7.1 needed here, updating doctrine/persistence to 1.3.6 will fix this issue.

greg0ire commented:
Should we close this PR?

sspat commented:
Yes, the issue is fixed by upgrading a dependency, no changes needed in the orm

cbichis commented:
Hi,

Looks like with 2.7.5 and doctrine/persistence at 1.3.8 this is still an issue.

beberlei commented:
@cbichis csn you be more specific please with exception message and full stack trace?

cbichis commented:
I am running Doctrine ORM 2.7.5, doctrine/persistence 1.3.8, doctrine/reflextion at 1.2.2.

$entity = new Supplier();
$this->form->bind($entity);
triggering:

File:
    /project/module/Application/src/Entity/Traits/IdTrait.php:18
Message:
    Typed property Supplier::$id must not be accessed before initialization
Stack trace:
    #0 /project/vendor/doctrine/doctrine-module/src/DoctrineModule/Stdlib/Hydrator/DoctrineObject.php(189): Supplier->getId()
    #1 /project/vendor/doctrine/doctrine-module/src/DoctrineModule/Stdlib/Hydrator/DoctrineObject.php(87): DoctrineModule\Stdlib\Hydrator\DoctrineObject->extractByValue()
    #2 /project/vendor/laminas/laminas-form/src/Fieldset.php(659): DoctrineModule\Stdlib\Hydrator\DoctrineObject->extract()
    #3 /project/vendor/laminas/laminas-form/src/Form.php(979): Laminas\Form\Fieldset->extract()
    #4 /project/vendor/laminas/laminas-form/src/Form.php(314): Laminas\Form\Form->extract()
    #5 /project/module/Application/src/Service/DoctrineCrudTrait.php(163): Laminas\Form\Form->bind()
And the entity has:
/**
 * @ORM\Id
 * @ORM\GeneratedValue(strategy="IDENTITY")
 * @ORM\Column(name="id", type="integer", nullable=false, options={"unsigned"=true})
 * @Annotation\Exclude()
 */
protected int $id;

beberlei commented:
This is a bug in Doctrine Laminas Module their Hydrator support, please open an issue on their repo with the code and stacktrace from your comment.

cbichis commented:
Thanks. Even after updating DoctrineModule/DoctrineORMModule the error stays so yes... I will open issue.
